src/payload.rs

```diff
@@ -0,0 +1,245 @@
+
+impl EsbHeader {
+    /// convert into a packed representation meant for internal
+    /// data queuing purposes
+    fn into_bytes(self) -> HeaderBytes {
+        HeaderBytes([
+            self.rssi,
+            self.pipe,
+            // DO NOT REORDER!
+            self.length,
+            self.pid_no_ack,
+        ])
+    }
+
+    /// convert from a packed representation
+    pub(crate) fn from_bytes(bytes: HeaderBytes) -> Self {
+        Self {
+            rssi: bytes.0[Self::rssi_idx()],
+            pipe: bytes.0[Self::pipe_idx()],
+            length: bytes.0[Self::length_idx()],
+            pid_no_ack: bytes.0[Self::pid_no_ack_idx()],
+        }
+    }
+
+    /// Accessor for the Pipe ID of the packet
+    pub fn pid(&self) -> u8 {
+        self.pid_no_ack >> 1
+    }
+
+    /// Accessor for the no-ack field of the packet
+    pub fn no_ack(&self) -> bool {
+        self.pid_no_ack & 1 != 1
+    }
+
+    /// Accessor for the length (in bytes) of the payload
+    pub fn payload_len(&self) -> usize {
+        usize::from(self.length)
+    }
+
+    /// Byte index of the RSSI field
+    const fn rssi_idx() -> usize {
+        0
+    }
+
+    /// Byte index of the pipe field
+    const fn pipe_idx() -> usize {
+        1
+    }
+
+    /// Byte index of the payload length field
+    const fn length_idx() -> usize {
+        // DO NOT CHANGE! HW DEPENDANT
+        2
+    }
+
+    /// Byte index of the pid_no_ack field
+    const fn pid_no_ack_idx() -> usize {
+        // DO NOT CHANGE! HW DEPENDANT
+        3
+    }
+
+    /// Size of the header (packed) in bytes
+    pub(crate) const fn header_size() -> usize {
+        core::mem::size_of::<HeaderBytes>()
+    }
+
+    /// Offset of the bytes needed for DMA processing
+    const fn dma_payload_offset() -> usize {
+        2
+    }
+}
+
+/// A handle representing a grant of a readable packet
+///
+/// This exposes the bytes of a payload that have either
+/// been sent FROM the app, and are being read by the RADIO,
+/// or a payload that has send FROM the Radio, and is being
+/// read by the app
+pub struct PayloadR<N: ArrayLength<u8>> {
+    grant: FrameGrantR<'static, N>,
+}
+
+impl<N> PayloadR<N>
+where
+    N: ArrayLength<u8>,
+{
+    /// Create a wrapped Payload Grant from a raw BBQueue Framed Grant
+    pub(crate) fn new(raw_grant: FrameGrantR<'static, N>) -> Self {
+        Self { grant: raw_grant }
+    }
+
+    /// Obtain a copy of the header encoded in the current grant
+    pub fn get_header(&self) -> EsbHeader {
+        const LEN: usize = EsbHeader::header_size();
+        let mut bytes = [0u8; LEN];
+        bytes.copy_from_slice(&self.grant[..LEN]);
+        EsbHeader::from_bytes(HeaderBytes(bytes))
+    }
+
+    /// Obtain a pointer to the data to provide to the RADIO DMA.
+    ///
+    /// This includes part of the header, as well as the full payload
+    pub(crate) fn dma_pointer(&self) -> *const u8 {
+        (&self.grant[EsbHeader::dma_payload_offset()..]).as_ptr()
```

Comment:
Also here, is there a reason you got rid of my `unsafe` stuff? I would guess this will get optimized down to the same thing, but it does technically make a new slice, when we don't need one (instead of just a pointer)

Reply:
I just thought it was an unnecessary `unsafe`, it seems pretty trivial to optimize so I guess we won't have problems, and it also prevents out of bounds access (although that would be pretty hard to get wrong in the previous code, anyway).